### PR TITLE
V0.12.0 prepare

### PR DESCRIPTION
**Pull request by alexstratos**

### CODE REVIEW

**Review by Xiong-stratos — APPROVED** [review]

**Comment by BoThe1K** [comment]
[QB-2374: Add web3 tx handling pr message (](https://github.com/stratosnet/stratos-chain/pull/349/commits/3108604cc5204db5c5f7955fd3e7471b40b9e3ab)https://github.com/stratosnet/stratos-chain/pull/334[)](https://github.com/stratosnet/stratos-chain/pull/349/commits/3108604cc5204db5c5f7955fd3e7471b40b9e3ab)

 Added an interface to better handling all transactions with `eth_getTransactionReceipt` rpc

[qb-2420: Fix tx and block search (](https://github.com/stratosnet/stratos-chain/pull/349/commits/b637e635ed9c55c6361ec3c5115ed43bd5d15374)https://github.com/stratosnet/stratos-chain/pull/342[)](https://github.com/stratosnet/stratos-chain/pull/349/commits/b637e635ed9c55c6361ec3c5115ed43bd5d15374)

We have applied our cometbft fork in order to process queries filter `http://localhost:26657/tx_search?query=%22stratos.evm.v1.EventEthereumTx.eth_hash=%270x6e4bcdf191b167bc362f622b98c4523189a67170fdde766baaf28aa13118b0a6%27%22` so we are able to query with TypedEvents.

[QB-2394: Fix EVM api](https://github.com/stratosnet/stratos-chain/pull/349/commits/483ccea6b05b1a61973f956425d69d36591e5e8d)

Updated cosmos mempool to support `MsgEthereumTx` transactions;
Added support for old transactions handling before v012 (mostly for readonly);

[QB-2410: Add contract check into ante](https://github.com/stratosnet/stratos-chain/pull/349/commits/7072dc4e2519920a58abbbea82e5762b6be032f2)

Added smart contract size check into ante handler

[QB-2394: Apply fix for a nw tx/block indexer](https://github.com/stratosnet/stratos-chain/pull/349/commits/381ed526e7250e8c8795a78b58cd7e003efdd609)

Fixed <v012 params obtaining through keystore

[QB-2372: implement proxy, add prepay (](https://github.com/stratosnet/stratos-chain/pull/349/commits/05f8e4bb775306f86f1e539f0e8bf4d9df10b4c5)https://github.com/stratosnet/stratos-chain/pull/337[)](https://github.com/stratosnet/stratos-chain/pull/349/commits/05f8e4bb775306f86f1e539f0e8bf4d9df10b4c5)

Returned back `opPrepay` opcode and make it available with v012;
Added migration for prepay proxy + implementation contract;
Added custom proposal mechanics to update implementation of the prepay proxy;
Added one more StateDB to support our `x` modules (in order to revert snapshots);

[Fix after events merge](https://github.com/stratosnet/stratos-chain/pull/349/commits/ddb1a96ccf1a061ad7ef1c2d35c71e5295ac93bc)
Removed events which is not used anymore

**Comment by Xiong-stratos** [comment]
[https://github.com/stratosnet/stratos-chain/pull/349/commits/eff1103a34e37893a473b98c136f78329a1b1185](https://github.com/stratosnet/stratos-chain/pull/349/commits/eff1103a34e37893a473b98c136f78329a1b1185)
Bump cosmos-sdk/ibc-go/tendermint/tm-db
cosmos-sdk -> v0.47.5
ibc-go     -> v7.3.0
tendermint -> cometbft v0.37.2
tm-db      -> cometbft-db v0.8.0

[https://github.com/stratosnet/stratos-chain/pull/349/commits/90f3ad658c0f507c29b4b6a595c85b80b053eafe](https://github.com/stratosnet/stratos-chain/pull/349/commits/90f3ad658c0f507c29b4b6a595c85b80b053eafe)
Update the store key, getter & setter of params in each modules due to cosmos-sdk manage params by x/consensus.
Update some properties in proto files.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/cd6e0ee3b92730830c999433b173a1fc19fed0d9](https://github.com/stratosnet/stratos-chain/pull/349/commits/cd6e0ee3b92730830c999433b173a1fc19fed0d9)
Use TypedEvent instead of deprecated Event

[https://github.com/stratosnet/stratos-chain/pull/349/commits/b2d6bfc59cb31893b030a73d5bd405a5ad891daa](https://github.com/stratosnet/stratos-chain/pull/349/commits/b2d6bfc59cb31893b030a73d5bd405a5ad891daa)
Add UpdateParams() for each modules.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/925b795f7cff474acad65ba0c074ad134f7963f5](https://github.com/stratosnet/stratos-chain/pull/349/commits/925b795f7cff474acad65ba0c074ad134f7963f5)
Add migration upgrade for params store.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/da6ce69d2a228eb2fcab6a933c14a12f179dfd44](https://github.com/stratosnet/stratos-chain/pull/349/commits/da6ce69d2a228eb2fcab6a933c14a12f179dfd44)
Remove files which is not used anymore
Update the implementation of tx fee anteHandler to support min-gas-price use different registered denoms(wei/gwei/stos)
Register upgrade handler.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/19d1774c3abd53af2a1ec06a19fe85592c2e71e0](https://github.com/stratosnet/stratos-chain/pull/349/commits/19d1774c3abd53af2a1ec06a19fe85592c2e71e0)
Add beneficiary address for meta nodes.
Update migration upgrade for meta node store change.
Reward distribution by beneficiary address instead of owner address.
Fixed some cli flag description error.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/442c4235bfe629af81d671855f778ce0cac2b91b](https://github.com/stratosnet/stratos-chain/pull/349/commits/442c4235bfe629af81d671855f778ce0cac2b91b)
Add query: TotalRewardByEpoch
Update migration upgrade for TotalReward (don't merge legacy data into store when update, query data from tx log for legacy epochs)

[https://github.com/stratosnet/stratos-chain/pull/349/commits/6de0a85f60beec449ca30109f62a6a485b6cb363](https://github.com/stratosnet/stratos-chain/pull/349/commits/6de0a85f60beec449ca30109f62a6a485b6cb363)
Fixed some unit tests.
Added bls signature test.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/d542725ec0383c53bb6f9bb16e86ee8afb00b1fa](https://github.com/stratosnet/stratos-chain/pull/349/commits/d542725ec0383c53bb6f9bb16e86ee8afb00b1fa)
Fix migration upgrade fail issue.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/8cc83ac748d7624f260364800e4552eac1da6db7](https://github.com/stratosnet/stratos-chain/pull/349/commits/8cc83ac748d7624f260364800e4552eac1da6db7)
Update VolumeReport query API:  when query epoch 0, return the latest epoch.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/4941432f869cc631e2fc3aa8d6d45fd1a8198b44](https://github.com/stratosnet/stratos-chain/pull/349/commits/4941432f869cc631e2fc3aa8d6d45fd1a8198b44)
Update protobuf scripts & generated API files.
Use proto marshaller to get bls sign bytes.
Bug fix: load total_mined_token from genesis file.
Total_mined_token migration update, add 1M stos which is failed to load from genesis file in previous version.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/4e19cff044cdc61c41f482ebd227930337df9e42](https://github.com/stratosnet/stratos-chain/pull/349/commits/4e19cff044cdc61c41f482ebd227930337df9e42)
Add beneficiary address for resource node.
Update migration upgrade for resource node store change.
Update api msgs.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/d81c5edbc743fba479b03a86cf42ed557ae09336](https://github.com/stratosnet/stratos-chain/pull/349/commits/d81c5edbc743fba479b03a86cf42ed557ae09336)
Remove useless EventMessage in each module.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/5d58329c9b40418c4b985c7d881db33fb2a40f3a](https://github.com/stratosnet/stratos-chain/pull/349/commits/5d58329c9b40418c4b985c7d881db33fb2a40f3a)
Remove: HandleMsgWithdrawMetaNodeRegistrationDeposit(MsgWithdrawMetaNodeRegistrationDeposit) returns (MsgWithdrawMetaNodeRegistrationDepositResponse);
Add: HandleMsgKickMetaNodeVote(MsgKickMetaNodeVote) returns (MsgKickMetaNodeVoteResponse);
Remove vote pool when 1/3 votes reject
Remove expired vote pool every 10000 blocks
Optimized some event attributes.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/f6cea5c437b65e323a8ebf1c27d862ea9c93cc61](https://github.com/stratosnet/stratos-chain/pull/349/commits/f6cea5c437b65e323a8ebf1c27d862ea9c93cc61)
Added missed gensis import/export store data.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/e4f3aba19daf72652ba39251d655537916c9e8f3](https://github.com/stratosnet/stratos-chain/pull/349/commits/e4f3aba19daf72652ba39251d655537916c9e8f3)
Allow suspended resource node to unbond.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/9a722c0f904e0bdb1d24c29424877176983907d3](https://github.com/stratosnet/stratos-chain/pull/349/commits/9a722c0f904e0bdb1d24c29424877176983907d3)
Rename some REST query API to follow the pattern of cosmos-sdk
Fixed some query API

[https://github.com/stratosnet/stratos-chain/pull/349/commits/8902811f4821016228263b488223a28f806a145c](https://github.com/stratosnet/stratos-chain/pull/349/commits/8902811f4821016228263b488223a28f806a145c)
Use cosmossdk.io/math.Int instead of github.com/cosmos/cosmos-sdk/types.Int

[https://github.com/stratosnet/stratos-chain/pull/349/commits/1abf5b08ef37971adc07fcd73d25fbf29f6e0104](https://github.com/stratosnet/stratos-chain/pull/349/commits/1abf5b08ef37971adc07fcd73d25fbf29f6e0104)
gRPC query API changes

[https://github.com/stratosnet/stratos-chain/pull/349/commits/a29695c76263cc6fd47b687d4af9122899d9c684](https://github.com/stratosnet/stratos-chain/pull/349/commits/a29695c76263cc6fd47b687d4af9122899d9c684)
gRPC query API changes and parameter change.

[https://github.com/stratosnet/stratos-chain/pull/349/commits/e2f3de2870901b28291ea46c8f93f1ddbd81151c](https://github.com/stratosnet/stratos-chain/pull/349/commits/e2f3de2870901b28291ea46c8f93f1ddbd81151c)
Add kickMetaNodeVotePool to genesis file

**Pull request merged**
